File: 2022/go/04.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Day4 struct{}

func (*Day4) Day() int {
	return 4
}

func (puzzle *Day4) Run(input string) {
	assignmentPairList := puzzle.createAssignmentPairList(input)

	puzzle.part1(assignmentPairList)
	puzzle.part2(assignmentPairList)
}

func (puzzle *Day4) part1(assignmentPairList []pair) {
	result := 0

	for _, pair := range assignmentPairList {
		if pair.lhs.isFullyContaining(pair.rhs) || pair.rhs.isFullyContaining(pair.lhs) {
			result++
		}
	}

	fmt.Printf("Part1: %d\n", result)
}

func (puzzle *Day4) part2(assignmentPairList []pair) {
	result := 0

	for _, pair := range assignmentPairList {
		if pair.lhs.isFullyContaining(pair.rhs) ||
			pair.rhs.isFullyContaining(pair.lhs) ||
			pair.lhs.isOverlapping(pair.rhs) {
			result++
		}
	}

	fmt.Printf("Part2: %d\n", result)
}

type assignment struct {
	start int
	end   int
}

type pair struct {
	lhs *assignment
	rhs *assignment
}

func (*Day4) newAssignment(pair string) *assignment {
	parts := strings.Split(strings.TrimSpace(pair), "-")

	start, _ := strconv.Atoi(strings.TrimSpace(parts[0]))
	end, _ := strconv.Atoi(strings.TrimSpace(parts[1]))

	return &assignment{
		start: start,
		end:   end,
	}
}

func (a *assignment) isFullyContaining(b *assignment) bool {
	return b.start >= a.start && b.end <= a.end
}

func (a *assignment) isOverlapping(b *assignment) bool {
	start := Max(a.start, b.start)
	end := Min(a.end, b.end)

	return start <= end
}

func (puzzle *Day4) createAssignmentPairList(input string) []pair {
	pairs := []pair{}

	for _, line := range strings.Split(input, "\n") {
		line = strings.TrimSpace(line)

		if line == "" {
			continue
		}

		parts := strings.Split(line, ",")

		pairs = append(pairs, pair{
			lhs: puzzle.newAssignment(parts[0]),
			rhs: puzzle.newAssignment(parts[1]),
		})
	}

	return pairs
}

File: 2022/cs/04.cs
namespace Day4Namespace;

public record struct Range(int Start, int End)
{
    public bool IsFullyContaining(Range other)
    {
        return other.Start >= Start && other.End <= End;
    }

    public bool IsOverlapping(Range other)
    {
        int start = Math.Max(Start, other.Start);
        int end = Math.Min(End, other.End);

        return start <= end;
    }

    public static Range Parse(string value)
    {
        string[] parts = value.Trim().Split('-');

        return new Range(
            int.Parse(parts[0].Trim()),
            int.Parse(parts[1].Trim())
        );
    }
}

class Day4 : IPuzzle
{
    public int Day => 4;
    public bool IsTest => false;

    public void Run(string input)
    {
        List<(Range lhs, Range rhs)> assignmentPairList = CreateAssignmentPairList(input);

        int result1 = 0;
        int result2 = 0;

        foreach ((Range lhs, Range rhs) in assignmentPairList)
        {
            bool isFullyContained = lhs.IsFullyContaining(rhs) || rhs.IsFullyContaining(lhs);

            if (isFullyContained)
                result1++;

            if (isFullyContained || lhs.IsOverlapping(rhs))
                result2++;
        }

        Console.WriteLine($"Part1: {result1}");
        Console.WriteLine($"Part2: {result2}");
    }

    private static List<(Range lhs, Range rhs)> CreateAssignmentPairList(string input)
    {
        var result = new List<(Range lhs, Range rhs)>();

        foreach (string line in input.Split('\n').Select(x => x.Trim()))
        {
            if (line == string.Empty)
                continue;

            string[] pairParts = line.Split(',');

            result.Add((Range.Parse(pairParts[0]), Range.Parse(pairParts[1])));
        }

        return result;
    }
}
